### DISCUSSION TITLE
Symfony Turbo Pusher

### DISCUSSION THREAD

**Discussion by EvilSpartans** [symfony/symfony#43511] [Q&A]
Hello ! Since installing Symfony UX: Turbo, I have a problem with Pusher. When I browse through multiple conversations and then send a message, it is duplicated by the number of conversations I clicked. There is only one saved message, but the user sees his message being sent in several copies.

Here is the **Javascript**:
```
  var pusher = new Pusher('ba75523bee28d7c644f2', {
        cluster: 'eu',
    });

    var channel = pusher.subscribe('my-channel-' + JSON.parse('{{ messaging.id | json_encode | raw }}'));
    var currentUser = JSON.parse('{{ app.user.id | json_encode | raw }}');

    channel.bind('my_event',
        function (data) {
            $('.lala').append('<li>'+ data['content'] +'</li >');
        });

    channel.bind('pusher:subscription_succeeded', function (members) { });
```
Here is the **controller** : 
```
/**
     * @Route("/new-message-{id}", name="newMessage", methods={"POST"})
     *
     * @throws PusherException|GuzzleException
     */
    public function message(Messaging $messaging, Request $request, PusherService $pusherService): Response
    {
        $user = $this->getUser();
        $pusherService->sendMessage();
        $message = new Message();

        if ($request->isMethod(Request::METHOD_POST) && $request->get('content') != null) {
            $message->setContent($request->get('content'));
            $entityManager = $this->getDoctrine()->getManager();
            $entityManager->persist($message);
            $entityManager->flush();

            $data['content'] = $message->getContent();
            $key = $messaging->getId();

            if ($pusherService->sendMessage()->trigger("my-channel-${key}", 'my_event', $data)) {
                echo 'success';
            } else {
                header('', true, 403);
                echo 'error';
            }
            return new JsonResponse(Response::HTTP_OK);
        }
    }
```
Thanks for your help !

**Comment by stof** [comment]
Without seeing *where* this JS code lives, we cannot really help.

When using Turbo, the whole page is not reloaded. So what might happen is that your page gets registered to pusher multiple times if that JS code executes multiple times.

**Reply by EvilSpartans** [reply]
What do you need exactly ? Here is the full project : https://github.com/EvilSpartans/Jobissim

**Reply by stof** [reply]
looking at the code, my guess was right. Your JS code is in the body of the page, and does not unregister its pusher subscriber when Turbo replaces the body. See https://turbo.hotwired.dev/handbook/building

This is why Symfony UX is also recommending the usage of Stimulus, which exposes a way to initialize a component, but also to clean it when it leaves the page.

**Reply by EvilSpartans** [reply]
I can see you are from Paris, I’m French so if you want you can speak French :) I can’t use a stimulus controller because of the twig variables :/ how can I unregister pusher subscriber when turbo replace the body ? With a turbo:before-cache eventListener : pusher.unsubscribe ?

I tried : 
```
    document.addEventListener("turbo:before-cache", function () {
        channel.unbind();
        pusher.unsubscribe('my-channel-' + JSON.parse('{{ messaging.id | json_encode | raw }}'));
    })
```
But not working :/

**Reply by EvilSpartans** [reply]
Can you help me to do it ? :/

**Reply by stof** [reply]
to me, the solution here is to rely on a data attribute or a `<meta>` tag to give some data from Twig to the HTML and then the JS, instead of injecting the value directly in the JS code from Twig. This way, it removes the need for the JS code itself to be inline (and so allows using stimulus or other tools that can handle detaching components). Trying to do that for inline JS code seems harder

**Reply by EvilSpartans** [reply]
OKay now I have the good stimulus **controller** : 
```
import { Controller } from "stimulus";

export default class extends Controller {
  static values = { userId: String, messageId: String };

  connect() {
    // Pusher.logToConsole = true;

    var pusher = new Pusher("ba75523bee28d7c644f2", {
      cluster: "eu",
    });
    var channel = pusher.subscribe("my-channel-" + this.messageIdValue);
    var currentUser = this.userIdValue;
    channel.bind("my_event", function (data) {
      console.log(data);
      $(".lala").append(
        data.author === currentUser
          ? '<li class = "chat-left"><div class="chat-avatar">' +
              (data["image"] !== null
                ? '<img src="/uploads/avatars/' + data["image"] + '">'
                : '<img src="/assets/images/nopic.png">') +
              '<div class="chat-name">' +
              data["user"] +
              '</div></div><div class="chat-text">' +
              data["content"] +
              '</div><div class="chat-hour">' +
              data["date"] +
              '<span class="fa fa-check-circle"></span></div></li>'
          : '<li class = "chat-right"><div class="chat-hour">' +
              data["date"] +
              '<span class="fa fa-check-circle"></span></div><div class="chat-text">' +
              data["content"] +
              '</div><div class="chat-avatar">' +
              (data["image"] !== null
                ? '<img src="/uploads/avatars/' + data["image"] + '">'
                : '<img src="/assets/images/nopic.png">') +
              '<div class="chat-name">' +
              data["user"] +
              "</div></div ></li > "
      );
    });

    channel.bind("pusher:subscription_succeeded", function (members) {
      console.log("successfully subscribed!");
    });
  }
}
```
and the **view** : 

```
<div class="lala" data-controller="pusher" data-pusher-user-id-value="{{ app.user.id }}"
        data-pusher-message-id-value="{{ messaging.id }}"> 
```
But messages are still duplicated ! :'(

**Reply by stof** [reply]
> But messages are still duplicated !

Well, you haven't implemented the logic for `disconnect()`, so this is normal...

**Reply by EvilSpartans** [reply]
Oh okay sorry, can you explain me how can I implement the logic for disconnect() please ? :)

**Reply by stof** [reply]
Well, probably something like `this.pusher = ...` in `connect()` (instead of using local variables) and then using `this.pusher.unsubscribe()` in `disconnect()`

**Reply by EvilSpartans** [reply]
I didn't really understand, but I tried this : 
```
  disconnect() {
    var channel = pusher.unsubscribe("my-channel-" + this.messageIdValue);
    channel.unbind();
  }
```
But not working for the moment

**Reply by stof** [reply]
Maybe you should ask for support on the pusher-js repo for the right way to unsubscribe, rather than asking it here. You have a better chance to get an answer as this is about their library.

**Reply by EvilSpartans** [reply]
Yes, but where can I ask this from this repo : https://github.com/pusher/pusher-js ? 
I think something like this should work : 
```
  disconnect() {
    var pusher = new Pusher("ba75523bee28d7c644f2", {
      cluster: "eu",
    });
    var channel = pusher.subscribe("my-channel-" + this.messageIdValue);
    pusher.connection.bind(
      "disconnected",
      pusher.unsubscribe("my-channel-" + this.messageIdValue),
      channel.unbind()
    );
  }
```
But still not working
